### PR TITLE
ADD: Add cmip6/cmip5 directory based parsers

### PR DESCRIPTION
**Pull request by mgrover1**
## Change Summary

Add directory-based parsers for CMIP 6/5 which are much faster than needing to open each file.

## Checklist

- [x] Unit tests for the changes exist
- [x] Tests pass on CI
- [x] Documentation reflects the changes where applicable

### CODE REVIEW

**Comment by mgrover1** [comment]
@andersy005 this is mostly there - just need to add a test for the CMIP5 parser

File: ecgtools/parsers/cmip.py [hunk 1]
```diff
@@ -86,3 +88,97 @@ def parse_cmip6(file):
...
+    filename_template = '{variable_id}_{table_id}_{source_id}_{experiment_id}_{member_id}_{grid_label}_{time_range}.nc'
+
+    gridspec_template = (
+        '{variable_id}_{table_id}_{source_id}_{experiment_id}_{member_id}_{grid_label}.nc'
+    )
+    templates = [filename_template, gridspec_template]
+    fileparts = reverse_filename_format(basename, templates=templates)
+    try:
+        parent = str(pathlib.Path(file).parent)
+        print(parent)
```
Comment (andersy005) [inline ecgtools/parsers/cmip.py:125]:
```suggestion
```

File: ecgtools/parsers/cmip.py [hunk 2]
```diff
@@ -86,3 +88,97 @@ def parse_cmip6(file):
...
+        fileparts['version'] = version
+        fileparts['path'] = file
+        if fileparts['member_id'].startswith('s'):
+            fileparts['dcpp_init_year'] = float(fileparts['member_id'].split('-')[0][1:])
+            fileparts['member_id'] = fileparts['member_id'].split('-')[-1]
+        else:
+            fileparts['dcpp_init_year'] = np.nan
+
+    except Exception:
+        pass
```
Comment (andersy005) [inline ecgtools/parsers/cmip.py:143]:
```suggestion
          return {INVALID_ASSET: file, TRACEBACK: traceback.format_exc()}
```

File: ecgtools/parsers/cmip.py [hunk 3]
```diff
@@ -86,3 +88,97 @@ def parse_cmip6(file):
...
+    fileparts['modeling_realm'] = realm
+    fileparts['version'] = version
+    fileparts['path'] = file
+    try:
+        part1, part2 = str(pathlib.Path(file).parent.split(fileparts['experiment']))
+        part1 = part1.strip('/').split('/')
+        fileparts['institute'] = part1[-2]
+        fileparts['product_id'] = part1[-3]
+    except Exception:
+        pass
```
Comment (andersy005) [inline ecgtools/parsers/cmip.py:182]:
```suggestion
          return {INVALID_ASSET: file, TRACEBACK: traceback.format_exc()}
```

File: ecgtools/parsers/utilities.py [hunk 4]
```diff
@@ -16,3 +23,53 @@ def extract_attr_with_regex(
...
+
+    print('Getting list of assets...\n')
+    filelist = [_file_dir_files(directory) for directory in dirs]
+    # watch progress
+    with ProgressBar():
+        filelist = dask.compute(*filelist)
+
+    filelist = list(itertools.chain(*filelist))
+    print('\nDone...\n')
+    return filelist
```
Comment (andersy005) [inline ecgtools/parsers/utilities.py:75]: Is this necessary? i thought this gets taken care of by the `Builder` class

Comment (mgrover1) [reply ecgtools/parsers/utilities.py:75]: Good point - it is not... we do not use it in the actual parsing function. I can go ahead and remove. Thanks for the other suggestions too!

**Comment by mgrover1** [comment]
@andersy005 found one more bug that was fixed the most recent commit, and addressed those comments + suggestions!

**Review by andersy005 — APPROVED** [review]
Thank you, @mgrover1!

**Pull request merged**